matrix_sdk_appservice/src/lib.rs

```diff
@@ -217,63 +205,94 @@ impl Appservice {
     ) -> Result<Self> {
         let homeserver_url = homeserver_url.try_into()?;
         let server_name = server_name.try_into()?;
+        let registration = Arc::new(registration);
+        let clients = Arc::new(DashMap::new());
 
-        let client_sender_localpart = Client::new(homeserver_url.clone())?;
-
-        client_session_with_login_restore(
-            &client_sender_localpart,
-            &registration,
-            registration.sender_localpart.as_ref(),
-            &server_name,
-        )
-        .await?;
-
-        Ok(Appservice { homeserver_url, server_name, registration, client_sender_localpart })
+        Ok(Appservice { homeserver_url, server_name, registration, clients })
     }
 
-    /// Get a [`Client`]
+    /// Create a [`Client`]
+    ///
+    /// Will create and return a [`Client`] that's configured to [assert the
+    /// identity] on all outgoing homeserver requests if `localpart` is
+    /// given. If not given the `Client` will use the main user associated
+    /// with this appservice, that is the `sender_localpart` in the
+    /// [`AppserviceRegistration`].
     ///
-    /// Will return a `Client` that's configured to [assert the identity] on all
-    /// outgoing homeserver requests if `localpart` is given. If not given
-    /// the `Client` will use the main user associated with this appservice,
-    /// that is the `sender_localpart` in the [`AppserviceRegistration`]
+    /// This method is a singleton that saves the client internally for re-use
+    /// based on the `localpart`.
     ///
     /// # Arguments
     ///
     /// * `localpart` - The localpart of the user we want assert our identity to
     ///
     /// [registration]: https://matrix.org/docs/spec/application_service/r0.1.2#registration
     /// [assert the identity]: https://matrix.org/docs/spec/application_service/r0.1.2#identity-assertion
-    pub async fn client(&self, localpart: Option<&str>) -> Result<Client> {
+    pub async fn client(&mut self, localpart: Option<&str>) -> Result<Client> {
+        let client = self.client_with_config(localpart, ClientConfig::default()).await?;
+
+        Ok(client)
+    }
+
+    /// Same as [`Self::client`] but with the ability to pass in a
+    /// [`ClientConfig`]
+    pub async fn client_with_config(
+        &mut self,
+        localpart: Option<&str>,
+        config: ClientConfig,
+    ) -> Result<Client> {
         let localpart = localpart.unwrap_or_else(|| self.registration.sender_localpart.as_ref());
```

Comment:
Should be deduplicated and re-use `get_cached_client` instead, but was hitting some Send + Sync roadblock

Reply:
Noted in the tracking issue